### PR TITLE
Bazel build

### LINKED ISSUE #58: Makefile improvement
- [ ] documentation in the file
- [ ] `.PHONY` statements
- [x] allow for specific testing targets with variables

### PR DESCRIPTION
**Pull request by chrislovecnm**
Adding bazel to this project

Features

- go compilation
- creation of the operator container without docker
- container images are managed by rules_docker now - @chrisseto do we need to use the ubi container?
- testing now includes several scripts under hack
- k8s object deploys
- generating crds
- code generation
- runs go tests
- runs e2e tests by starting kind
- fully hermetic, with the exception of python3
- fully deterministic

TODO

This is a checklist for me to remember what I need to complete on this PR.

- [x] crd version is not getting included - no idea how to fix this.  I will open an issue. #130
- [x] headers are not added to crds
- [x] hack/bin/BUILD.bazel has dead code
- [x] hack/build/print-workspace-status.sh has incorrect registry. @chrisseto where are we putting this container? #127
- [x] file headers need updating
- [x] create an update-all.sh and create a make target for it
- [x] pkg/testutil/exec/exec.go needs some code clean-up
- [x] pkg/testutil/paths/paths.go
- [x] e2e testing is busted.  @chrisseto we need to talk about this.  The files are getting us into trouble. - not going to do in this PR.  Will open an issue #125
- [x] more testing
- [x] more code comments around bazel
- [x] fix k8s targets to have dynamic values

Not included
- we need to figure out checking headers with python. cc @chrisseto 
- we need to figure out python3 issues on the testing server
- tagging override for the release process.  Need to figure out stamping.

Closes #58 #55 #52 #48 #126 #124 https://github.com/cockroachdb/cockroach-operator/issues/128

Openned various issues including #129 which is migrating CI to bazel

### CODE REVIEW

File: config/crd/bases/crdb.cockroachlabs.com_crdbclusters.yaml [hunk 1]
```diff
@@ -1,23 +1,10 @@
...
-# See the License for the specific language governing permissions and
-# limitations under the License.
 
 ---
 apiVersion: apiextensions.k8s.io/v1beta1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.2.9
+    controller-gen.kubebuilder.io/version: (unknown)
```
Comment (chrislovecnm) [inline config/crd/bases/crdb.cockroachlabs.com_crdbclusters.yaml:7]: TODO - need to figure this out

File: config/rbac/role.yaml [hunk 2]
```diff
@@ -1,16 +1,3 @@
-# Copyright 2020 The Cockroach Authors
```
Comment (chrislovecnm) [inline config/rbac/role.yaml:1]: TODO - need to get headers added back

File: hack/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+sh_binary(
+    name = "update-deps",
+    srcs = ["update-deps.sh"],
+    args = [
+        "$(location %s)" % GO,
+        "$(location %s)" % GAZELLE,
+        "$(location %s)" % KAZEL,
+        "$(location %s)" % JQ,
+        "$(location :update-bazel)",
+        # "$(location :update-deps-licenses)",
```
Comment (chrislovecnm) [inline hack/BUILD.bazel:26]: TODO - I wonder if we should pull in these to the project as well

File: hack/bin/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    srcs = select({
+        ":darwin": ["@jq_osx//file"],
+        ":k8": ["@jq_linux//file"],
+    }),
+    outs = ["jq"],
+    cmd = "cp $(SRCS) $@",
+    visibility = ["//visibility:public"],
+)
+
+#genrule(
```
Comment (chrislovecnm) [inline hack/bin/BUILD.bazel:16]: TODO remove dead code

File: hack/build/print-workspace-status.sh [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+cat <<EOF
+STABLE_BUILD_GIT_COMMIT ${KUBE_GIT_COMMIT-}
+STABLE_BUILD_SCM_STATUS ${KUBE_GIT_TREE_STATE-}
+STABLE_BUILD_SCM_REVISION ${KUBE_GIT_VERSION-}
+STABLE_BUILD_MAJOR_VERSION ${KUBE_GIT_MAJOR-}
+STABLE_BUILD_MINOR_VERSION ${KUBE_GIT_MINOR-}
+STABLE_DOCKER_TAG ${APP_VERSION:-${KUBE_GIT_VERSION/+/-}}
+STABLE_DOCKER_REGISTRY ${DOCKER_REGISTRY:-us.gcr.io/chris-love-operator-playground}
+STABLE_DOCKER_PUSH_REGISTRY ${DOCKER_PUSH_REGISTRY:-${DOCKER_REGISTRY:-us.gcr.io/chris-love-operator-playground}}
```
Comment (chrislovecnm) [inline hack/build/print-workspace-status.sh:43]: TODO update this with the correct registry information

File: hack/build/version.sh [hunk 6]
```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/env bash
+
+# Copyright 2020 The Jetstack cert-manager contributors.
```
Comment (chrislovecnm) [inline hack/build/version.sh:3]: TODO headers

File: hack/update-crds.sh [hunk 7]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
```
Comment (chrislovecnm) [inline hack/update-crds.sh:1]: TODO create an update-all and create a make target for it

File: pkg/testutil/exec/exec.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"os"
+
+	"sigs.k8s.io/kubetest2/pkg/process"
+)
+
+// TODO document
+
+func StopKubeTest2(clusterName string) error {
```
Comment (chrislovecnm) [inline pkg/testutil/exec/exec.go:27]: TODO clean up code and add comments

File: pkg/testutil/paths/paths.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+package paths
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+func MaybeSetEnv(key, bin string, path ...string) {
```
Comment (chrislovecnm) [inline pkg/testutil/paths/paths.go:25]: TODO code comments

**Comment by chrislovecnm** [comment]
@chrisseto PTAL.  Also, CI is busted again.

File: BUILD.bazel [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+container_push(
+   name = "push_operator_image",
+   image = "//cmd/cockroach-operator:operator_image",
+   format = "Docker",
+   registry = "{STABLE_DOCKER_REGISTRY}",
+   repository = "cockroach-operator",
+   # TODO need a way to override this
+   # probably https://github.com/jetstack/cert-manager/blob/ba7871d5f685e0f8d2b256f3a1230f8786d110fe/build/BUILD.bazel#L58
+   tag = "{STABLE_BUILD_GIT_COMMIT}",
```
Comment (chrislovecnm) [inline BUILD.bazel:42]: TODO need tagging override for releases

**Comment by chrislovecnm** [comment]
@chrisseto PTAL

File: config/crd/bases/crdb.cockroachlabs.com_crdbclusters.yaml (hunk 1 above)
Comment (chrislovecnm) [reply config/crd/bases/crdb.cockroachlabs.com_crdbclusters.yaml:20]: Filed an issue in the backlog

File: hack/BUILD.bazel (hunk 3 above)
Comment (chrislovecnm) [reply hack/BUILD.bazel:28]: Low priority

File: hack/build/print-workspace-status.sh (hunk 5 above)
Comment (chrislovecnm) [reply hack/build/print-workspace-status.sh:43]: Issue filed

File: hack/update-crds.sh (hunk 7 above)
Comment (chrislovecnm) [reply hack/update-crds.sh:1]: No needed

File: pkg/testutil/exec/exec.go (hunk 8 above)
Comment (chrislovecnm) [reply pkg/testutil/exec/exec.go:26]: Done

File: pkg/testutil/paths/paths.go (hunk 9 above)
Comment (chrislovecnm) [reply pkg/testutil/paths/paths.go:27]: Done

File: Makefile.bazel [hunk 11]
```diff
@@ -0,0 +1,89 @@
+# Copyright 2020 The Cockroach Authors
```
Comment (chrisseto) [inline Makefile.bazel:1]: Should this become just `Makefile`?

File: WORKSPACE [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+)
+
+load("@io_bazel_rules_go//go:deps.bzl", "go_register_toolchains", "go_rules_dependencies")
+
+go_rules_dependencies()
+
+go_register_toolchains(
+    go_version = "1.14.4",
+    # todo nogo is failing
+    # nogo = "@//hack/build:nogo_vet",
```
Comment (chrisseto) [inline WORKSPACE:35]: Ran into the same issues in CC, stumbled across these:

https://github.com/bazelbuild/rules_go/issues/2513
https://github.com/bazelbuild/rules_go/issues/2396

File: api/v1alpha1/BUILD.bazel [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+go_test(
+    name = "go_default_test",
+    srcs = [
+        "cluster_defaults_test.go",
+        "cluster_types_test.go",
+        "suite_test.go",
+        "volume_test.go",
+    ],
+    # TODO use these values in the unit test instead of the hardcoded values
```
Comment (chrisseto) [inline api/v1alpha1/BUILD.bazel:38]: Is there an easy way to inject these values into `$PATH`?

File: Makefile.bazel (hunk 11 above)
Comment (chrislovecnm) [reply Makefile.bazel:1]: Yes.  Do you want to change CI now?

File: api/v1alpha1/BUILD.bazel (hunk 13 above)
Comment (chrislovecnm) [reply api/v1alpha1/BUILD.bazel:38]: So we have options.  

I am using https://github.com/cockroachdb/cockroach-operator/blob/5928ca90d499d672ac54fa68dcbecb7e3748849e/pkg/testutil/paths/paths.go#L40 to modify the `PATH`.  But I have the names of the binaries hardcoded in the unit test here:
https://github.com/cockroachdb/cockroach-operator/blob/5928ca90d499d672ac54fa68dcbecb7e3748849e/api/v1alpha1/suite_test.go#L63

My thought was to pass in the names of the binaries as parameters. 
The reason I have the test adjust the `PATH` is that only the `PATH` for this test suite is modified..

The other option is changing the PATH for bazel.  Which is messy, to say the least.

File: WORKSPACE (hunk 12 above)
Comment (chrislovecnm) [reply WORKSPACE:35]: So can this work?

**Pull request merged**